### PR TITLE
mgr/cephadm: don't cleanup the daemon keyring on failed redeploy

### PR DESCRIPTION
**Pull request by melissa-kun-li**
Fixes: https://tracker.ceph.com/issues/49872
Signed-off-by: Melissa Li <li.melissa.kun@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by melissa-kun-li** [comment]
make check failed:
```
unittest-seastore: ../src/crimson/os/seastore/onode_manager/staged-fltree/node_delta_recorder.h:21: virtual crimson::os::seastore::onode::DeltaRecorder::~DeltaRecorder(): Assertion `is_empty()' failed.
Aborting on shard 0. 
```

I think it's related to: https://tracker.ceph.com/issues/50028 ?

**Comment by adk3798** [comment]
> make check failed:
> 
> ```
> unittest-seastore: ../src/crimson/os/seastore/onode_manager/staged-fltree/node_delta_recorder.h:21: virtual crimson::os::seastore::onode::DeltaRecorder::~DeltaRecorder(): Assertion `is_empty()' failed.
> Aborting on shard 0. 
> ```
> 
> I think it's related to: https://tracker.ceph.com/issues/50028 ?

Yeah, that failure definitely looks unrelated to the changes in this pull request. There a few tests that will occasionally fail in make check (rbd unit tests for example) and sometimes make check will fail due to things unrelated to ceph (such as possibly failing to install some dependency). If you're confident, as you would be in this case, the failure is unrelated to your changes, you can just tell Jenkins to rerun the make check by commenting "jenkins test make check" on the pull request. We do it fairly often. As long as the make check isn't just repeatedly failing it isn't an issue.

**Comment by adk3798** [comment]
jenkins test make check

**Review by adk3798 — APPROVED** [review]
LGTM

**Comment by melissa-kun-li** [comment]
Thanks for letting me know @adk3798 !

**Comment by melissa-kun-li** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1701,7 +1701,9 @@ def _daemon_action(self,
             if daemon_spec.daemon_type != 'osd':
                 daemon_spec = self.cephadm_services[daemon_type_to_service(
                     daemon_spec.daemon_type)].prepare_create(daemon_spec)
-            return CephadmServe(self)._create_daemon(daemon_spec, reconfig=(action == 'reconfig'))
+            if action == 'reconfig':
+                return CephadmServe(self)._create_daemon(daemon_spec, reconfig=True)
+            return CephadmServe(self)._create_daemon(daemon_spec, redeploy=True)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1706]: Idempotency. `CephadmServe._create_daemon` is supposed to be idempotent. That's why we don't pass `redeploy` to it. I don't think we should pass redeploy here.

File: src/pybind/mgr/cephadm/serve.py [hunk 2]
```diff
@@ -918,7 +919,7 @@ def _create_daemon(self,
                         daemon_spec.name(), OrchestratorEvent.ERROR, f'Failed to {what}: {err}')
                 return msg
             except OrchestratorError:
-                if not reconfig:
+                if not reconfig and not redeploy:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/serve.py:922]: can we compute redeploy here? 

```python
redeploy = daemon_spec.name() in self.mgr.cache.get_daemon_names()
```

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (melissa-kun-li) [reply src/pybind/mgr/cephadm/module.py]: Changed it back to original

**Pull request merged**